### ISSUE TITLE
Try other memory-reduction strategies

### ISSUE THREAD

**Issue by jder** [m2lines/Samudra#444]
- [x] Profile current quarter-degree
- [x] Micro batching (#476)
- [ ] Activation offloading 
- [X] Disconnect gradients (#428)
- [ ] See if we can get rid of or merge slice_backward, pad, and `add` and `to` allocations at peak memory
- [x] #357 
- [ ] https://github.com/Open-Athena/Ocean_Emulator/issues/457
- [ ] reduce data loader duplication between steps in autoregressive rollout (one batch is 10% of dataset…)
- [ ] Hist = 0?

**Comment by alxmrs** [comment]
Maybe #433 is related to this? TBD, need to do more research on torch internals.

**Comment by jder** [comment]
ok, I ran on an 80 GB H100 on `45dc242c7667b`:

```
uv run python -m ocean_emulators.train configs/train_om4_halfdeg.yaml --experiment.data_root=/home/jrusak/orcd/pool/data/om4_sample_quarterdeg/ --data=@configs/data/public.yaml --train_time.start=2013-01-01 --train_time.end=2018-12-31 --val_time.start=2019-01-01 --val_time.end=2022-12-29  --experiment.wandb.mode=disabled --preemptible=false --experiment.prognostic_vars_key=thermo_dynamic_5 --experiment.boundary_vars=tau_hfds --batch_size=2 --steps='[1]' --profiler.cuda_snapshot_frequency=5
```

and got the following dump after 15 steps.

[cuda_mem_h100_quarter_top_5.json](https://github.com/user-attachments/files/23552797/cuda_mem_h100_quarter_top_5.json)

This is what the model looks like: https://gist.github.com/jder/11570394d6e5a0a0aba1d2170eed23fe. Interestingly the asymmetry in the Unet (final block is has more channels but same resolution as input) is what causes a lion's share of memory usage.

Perhaps unsurprisingly the activations vastly dominate memory usage over the input data. Also interesting that the input data seems to live for 2 steps instead of 1 which seems somewhat surprising. I think I found and fixed this bug (of mine) in #463.

**Comment by jder** [comment]
ok, looks like peak memory is at final unet block in backward pass, with a stack of activations from conv2d, batch norm, padding, gelu, and truncate (latter is from our capped gelu activation function). The behavior when you checkpoint some outer block seems to be that the entire interior is recomputed in the backward pass and *all* of that is in memory at once. There doesn't seem to be a way to prevent that, meaning that (at least for this last block) using *less* checkpointing (ie keeping around some but not all intermediate results in that block) would probably save peak memory. 

We also probably should not need separate allocations for all these things at all? Some of these are trivial transformations and seem like we should be able to avoid saving the intermediate outputs (perhaps with some form of kernel fusion).

**Comment by jder** [comment]
Ok, I tried torch.compile and it improves speed by 25%ish but actually increases peak memory usage slightly. There are many options we could try here but it does not seem super worthwhile at the moment given it has various drawbacks (requires a python development environment, does compilation at runtime, takes 15 minutes to start)

**Comment by jder** [comment]
See #481

**Comment by jder** [comment]
To reduce batch size, we'll probably want to move to another norm, see #405
